AB-3716 Sync banks & clearing numbers with Bankgirot

alexteg-advisa opened the pull request:
This aims to synchronize the clearing number and account number validation with the latest version (`2019-10-22`) of the ["Bank Account Numbers in Swedish Banks" manual](https://www.bankgirot.se/globalassets/dokument/anvandarmanualer/bankernaskontonummeruppbyggnad_anvandarmanual_sv.pdf) from Bankgirot.

It **adds** `Klarna Bank` & `MedMera Bank`, **removes** the `Royal Bank of Scotland` and **updates** the clearing number ranges for `Riksgälden` and `Swedbank`. It also **renames** `Amfa Bank` to `Svea Bank` and improves the spelling of some other bank names.

This also adds CI through [Github Actions](https://github.com/features/actions), since that was quick and simple to do and will help ensure all tests pass for PRs. As part of this the `master` branch has been protected to require a PR, review, successful tests & commit message linting before merges to `master`.

AB-3716

alexteg-advisa committed:
Remove Royal Bank of Scotland not operating abroad

The Royal Bank of Scotland has ceased all abroad operations, which
resulted in them ceasing all operations in Sweden in August 2015:
https://www.rbs.com/rbs/about/worldwide-locations/rbs-international-branch-network/sweden.html

Since a Royal Bank of Scotland account number was used in a test, the
test needs to be updated with a new account number.

AB-3716

alexteg-advisa committed:
Extend Swedbank clearing number range

The "Bank Account Numbers in Swedish Banks" manual from Bankgirot was
updated 2019-10-22 with some changes in which clearing number ranges
belong to Swedbank:
https://www.bankgirot.se/globalassets/dokument/anvandarmanualer/bankernaskontonummeruppbyggnad_anvandarmanual_sv.pdf

Swedbank has extended their 9300-9329 range to also include 9330-9349,
meaning that range now spans from 9300 to 9349.

AB-3716

alexteg-advisa committed:
Correct spelling of some bank names

Ensure bank names are consistent with their branding and Bankgirot
manual "Bank Account Numbers in Swedish Banks" as of version 2019-10-22:
https://www.bankgirot.se/globalassets/dokument/anvandarmanualer/bankernaskontonummeruppbyggnad_anvandarmanual_sv.pdf

AB-3716

alexteg-advisa committed:
Add new Klarna & MedMera banks

Klarna got their banking license on 19 June 2017:
https://sverigesradio.se/sida/artikel.aspx?artikel=6720487

MedMera Bank has also recently been cleared for clearing by Bankgirot
according to their manual "Bank Account Numbers in Swedish Banks" as of
2019-10-22:
https://www.bankgirot.se/globalassets/dokument/anvandarmanualer/bankernaskontonummeruppbyggnad_anvandarmanual_sv.pdf

AB-3716

alexteg-advisa committed:
Rename "Amfa Bank" to "Svea Bank"

Svea Ekonomi bought Amfa Bank on 10th November 2016, which resulted in
an official rebranding to "Svea Bank" on 20 November 2017. See press
release:
https://www.mynewsdesk.com/se/pressreleases/amfa-bank-byter-namn-till-svea-bank-2282938

AB-3716

alexteg-advisa committed:
Remove README demo link to expired domain

The jop.io domain seems to be up for sale at GoDaddy and doesn't host
the linked page anymore, so it's removed from the README.

AB-3716

alexteg-advisa committed:
Add Riksgälden clearing number range 9890-9899

Riksgälden has added a new clearing number range between 9890-9899,
which as opposed to their earlier range should be validated with mod10.

Riksgälden has since 2012 had an initiative to be part of the Bankgiro
clearing system to provide more efficient money transfers between
governmental agencies in the State's internal bank:
https://www.riksgalden.se/globalassets/dokument_sve/sib/ovriga/dataclearinganslutning.pdf

AB-3716

alexteg-advisa committed:
Remove superfluous comment listing supported banks

The list of banks already exists as a quite readable array in the same
file and nicely documented in the README as a list. No need to duplicate
it a third time - it just adds unnecessary maintenance.

AB-3716

alexteg-advisa commented:
I pushed some fixups that update the documentation with the added, removed & renamed banks.

simonbergstrom commented on kontonummer.js:
```diff
@@ -33,7 +24,7 @@
+        name    : 'Svea Bank',
         regex   : /^(966[0-9])/,
         modulo  : 11,
         lengths : {
```
Shouldn't control be 11 for this bank according to the new docs? 

Comment 2: "Checksum calculation is made on the entire Clearing number, and seven digits of the actual account number." --> 4 + 7 = 11

alexteg-advisa replied:
You are correct. I missed that this was more than just a renaming of a bank.

alexteg-advisa replied:
I've fixed this in the fixup commit c9469c8a966b939e59f89a7288bdd92a0de6b614

simonbergstrom commented on kontonummer.js:
```diff
@@ -267,13 +276,13 @@
-        name    : 'Royal Bank of Scotland',
-        regex   : /^(909[0-9])/,
+        name    : 'Riksgälden',
+        regex   : /^(988[0-9])/,
```
For this clearing number I think the control number should be 11 according to the docs. Riksgälden exist for two different interval of clearing numbers.

alexteg-advisa replied:
You are correct. I got confused by what `control` means in this context.

alexteg-advisa replied:
I've fixed this in the fixup commit feeaf554d2f156f6c1a3bacf43f44653ec7ba04b

simonbergstrom commented on kontonummer.js:
```diff
@@ -203,6 +203,15 @@
+        lengths : {
+            clearing : 4,
+            account  : 7,
+            control  : 10
```
Same here with control number. Comment 2 with full clearing + account number as control

alexteg-advisa replied:
You are totally correct. I got confused by what `control` means in this context.

alexteg-advisa replied:
I've fixed this in the fixup commit 2bb6cd2b5d745511c2469864d4e5f03634ef0971

simonbergstrom commented on kontonummer.js:
```diff
@@ -203,6 +203,15 @@
             account  : 7,
             control  : 10
         }
```
Länsförsäkringar for clearing: 9060-9069 is missing in our config. Should it be added?

alexteg-advisa replied:
Isn't Länsförsäkringar `9060-9069` covered by `906[0-9]` part of the regex in:

https://github.com/Advisa/kontonummer.js/blob/3fb909fff52c414deec744b59f595cef315b9e77/kontonummer.js#L180-L181

?

simonbergstrom replied:
Aah sorry you are totally correct. 👍

simonbergstrom commented:
Went through all banks in our configuration and compared it to the doc and came up with some comments. We can maybe check them together at some point to see if I have understand it correctly.

alexteg-advisa committed:
Group account number formats as in Bankgirot docs

Rather than repeating the lengths configuration for each bank, opening
up for mistakes, group the different types of account number lengths
in a constant that directly maps to the two main types of account
numbers and their comments in the Bankgirot documentation:
https://www.bankgirot.se/globalassets/dokument/anvandarmanualer/bankernaskontonummeruppbyggnad_anvandarmanual_sv.pdf

This results in 5 distinct sets of clearing number, account number &
control length combinations.

Definition of length keys:
- clearing: length of clearing number
- account: length of account number without clearing number
- control: number of last digits in whole account number
  (`clearing + account`) for mod 10/11 checksum calculation

This means the total account number length is `clearing + account` of
which the last `control` number of digits are used for the mod 10/11
checksum calculation.

AB-3716

alexteg-advisa commented on kontonummer.js:
```diff
@@ -158,6 +158,15 @@
+        lengths : {
+            clearing : 4,
+            account  : 7,
+            control  : 11
```
Sorry for misleading you, but `control` actually refers to the `x` last digits of the _combined_ clearing & account number that should be part of the mod 10/11 checksum calculation, respectively:

### Type 1 account numbers
Comment number:
1. "Clearing number with the exception of the first digit, and seven digits of the actual account number." (basically means the last 10 digits)
2. "Checksum calculation is made on the entire Clearing number, and seven digits of the actual account number." (basically means all the 11 digits)

### Type 2 account numbers
Comment number:
1. "Checksum calculation uses the last ten [10] digits using the modulus 10 check"
2. "Checksum calculation uses the 9 last digits using the modulus 11 check."
3. "Checksum calculation uses the last ten [10] digits using the modulus 10 check, according format for account number (clearing number not included)"

as defined in https://www.bankgirot.se/globalassets/dokument/anvandarmanualer/bankernaskontonummeruppbyggnad_anvandarmanual_sv.pdf

alexteg-advisa commented:
@simonbergstrom @johanadvisa Since both I & Johan got confused regarding what `control` means, I've simplified the length definitions by grouping the 2 different account number types types with a total of 5 variations expressed as comments in the [Bankgirot documentation](https://www.bankgirot.se/globalassets/dokument/anvandarmanualer/bankernaskontonummeruppbyggnad_anvandarmanual_sv.pdf) as a constant that just references the type & comment, rather than repeating the same length object.

This makes the code more DRY, makes it more difficult to makes mistakes and maps more clearly to the [Bankgirot documentation](https://www.bankgirot.se/globalassets/dokument/anvandarmanualer/bankernaskontonummeruppbyggnad_anvandarmanual_sv.pdf). See commit: 45177c970c6ce83cafc08c32e97a23ad3c36ada8 `Sync banks & clearing numbers with Bankgirot`. This is similar to how the original repo we forked cleaned up the repeated definitions: https://github.com/jop-io/kontonummer.js/commit/e7abc137343498d2b1264a52c4830ebedd59400c

alexteg-advisa commented:
Github actions CI running of tests has been split out to a separate PR #5 to enable #4 to be merged without this.

alexteg-advisa commented:
The commit "Group account number formats as in Bankgirot docs" (A) is already before the commit "Add new Klarna & MedMera banks" (B), it's just that I created the grouping commit A after B, but moved commit A before B in an interactive rebase and [Github orders commit by author date instead of topologically](https://github.com/isaacs/github/issues/386).

Output of `git log --oneline --graph --no-decorate` (which orders topologically) after squashing the fixup commits:
<img width="440" alt="Screenshot 2020-01-09 at 12 48 31" src="https://user-images.githubusercontent.com/18527849/72065604-e8516b80-32de-11ea-9ee3-02543fbe3af4.png">
